### ISSUE TITLE
File timestamp set to Jan 1, 1970

### ISSUE THREAD

**Issue by h8000** [spotify/docker-maven-plugin#58]
The timestamps of the files copied into the docker images is always Jan 1, 1970

**Comment by davidxia** [comment]
It was done in this commit https://github.com/spotify/docker-maven-plugin/commit/4970d4b10b7548cee853c869aa92892ad9d7a397. The reason was "Set mtime of copied files to 0 to enable Docker cache". Does this not work with something you're doing?

**Comment by sbrosinski** [comment]
This is causing a very "interesting" bug for me:

I'm using docker-maven-plugin to build a docker container for a spring-boot application. Upon copying my jar to ./target/docker and setting its timestamp to Jan 1, 1970, which is 0 in unix epoch time, Spring MVC fails with:

There was an unexpected error (type=Internal Server Error, status=500).
class path resource [public/index.html] cannot be resolved in the file system for resolving its last-modified timestamp

They're checking for the last-modified timestamp in org.springframework.core.io.AbstractResource:

```java
public long lastModified() throws IOException {
		long lastModified = getFileForLastModifiedCheck().lastModified();
		if (lastModified == 0L) {
			throw new FileNotFoundException(getDescription() +
					" cannot be resolved in the file system for resolving its last-modified timestamp");
		}
		return lastModified;
	}
```

So since you changed the timestamp to 0, they'll fail with a FileNotFoundException. This cost me a day of work and some grey hair. :-)

**Comment by davidxia** [comment]
@sbrosinski Aw man, apologies. Did you find a workaround? I'm not familiar with spring-boot.

**Comment by sbrosinski** [comment]
@davidxia no worries. My workaround for now is to have the docker container launch a script touching the jar first instead of using ENTRYPOINT in the Dockerfile to start the jar, e.g.:

#!/usr/bin/env bash
touch spring-boot.jar
java -jar spring-boot.jar

**Comment by davidxia** [comment]
@sbrosinski Yea i was thinking the same thing. Setting the timestamp to 1 here https://github.com/spotify/docker-maven-plugin/commit/4970d4b10b7548cee853c869aa92892ad9d7a397#diff-f381e322663b27c7b711e0c8768b91d6R568

**Comment by sbrosinski** [comment]
If you dig deeper and look at the JavaDoc for java.io.File.lastModified() it says:

A long value representing the time the file was
last modified, measured in milliseconds since the epoch
(00:00:00 GMT, January 1, 1970), or 0L if the
file does not exist or if an I/O error occurs

So I assume any Java code using java.io.File might run into trouble here.

**Comment by sbrosinski** [comment]
Great, thanks. And let me also thank you for doing such a good job on the maven plugin. It saved me quite some trouble here when we had to dockerize our Java apps. We're using it to create and push docker images from Jenkins to our registry.

**Comment by chrisdostert** [comment]
using release 0.2.8 and still not seeing this working even though the commit history shows it should be in as of 0.2.7?

**Comment by creactiviti** [comment]
Seeing the same issue with 0.2.8. Thanks for the workaround @sbrosinski. Great plugin otherwise.

**Comment by sjardine** [comment]
Still seeing this issue in 0.2.8.  I would suggest re-opening this issue.  Also, why would we set the mtime to 1 instead of the real mtime of the file copying?

**Comment by davidxia** [comment]
@sjardine The reason is here https://github.com/spotify/docker-maven-plugin/pull/67 which @sbrosinski reported https://github.com/spotify/docker-maven-plugin/issues/58#issuecomment-91189781.

**Comment by davidxia** [comment]
@sjardine @creactiviti @chrisdostert How is this bug manifesting for you, ie how to repro?

So it seems docker-maven-plugin 0.2.8 doesn't work. Does @sbrosinski's workaround in https://github.com/spotify/docker-maven-plugin/issues/58#issuecomment-91189781 work for everyone? Does upgrading to 0.2.9 do anything?

**Comment by sjardine** [comment]
I am thinking that 0.2.9 is working.  I will do some more testing to verify.

**Comment by jleskovar** [comment]
Just to clarify, the bug is still present in docker-maven-plugin 0.2.9, where the file mtime is set to 1. This can be easily verified by building an image using docker-maven-plugin, running the resulting container and then performing:

    docker exec <new container ID> stat /my-artifact.jar

For me, I get the following:

    docker exec high_yonath stat /application-0.0.1-SNAPSHOT.jar
      File: ‘/application-0.0.1-SNAPSHOT.jar’
      Size: 43655824  	Blocks: 85280      IO Block: 4096   regular file
    Device: 38h/56d	Inode: 53          Links: 1
    Access: (0664/-rw-rw-r--)  Uid: (    0/    root)   Gid: (    0/    root)
    Access: 2015-06-09 02:42:47.041935346 +0000
    Modify: 1970-01-01 00:00:00.000000000 +0000
    Change: 2015-06-09 02:42:47.065935346 +0000
     Birth: -

This results in me seeing the problem @sbrosinski sees, as I'm also using this for a Spring Boot application.

**Comment by sjardine** [comment]
Yeah....this is still broken in 0.2.9.  I have the same question as @jleskovar at the end of his post.  For now, I have created my own build of the plugin with the setModifiedTime line removed.

**Comment by davidxia** [comment]
@sjardine @jleskovar Thanks for testing again. The reason for fixing the mtime is to let Docker cache the files if their contents haven't changed during `docker build`. This makes `docker build` faster since it can reuse layers.

From [Docker docs](https://docs.docker.com/articles/dockerfile_best-practices/#build-cache):

> During the process of building an image Docker will step through the instructions in your Dockerfile executing each in the order specified. As each instruction is examined Docker will look for an existing image in its cache that it can reuse, rather than creating a new (duplicate) image
> ...
> In the case of the ADD and COPY instructions, the contents of the file(s) being put into the image are examined. Specifically, a checksum is done of the file(s) and then that checksum is used during the cache lookup. If anything has changed in the file(s), **including its metadata**, then the cache is invalidated.

I'd like to keep the fixed mtime if we can fix this issue. Maybe fixing the mtime to epoch + 1 day will work for you? Please test out https://github.com/spotify/docker-maven-plugin/pull/98 and let me know.

When I `stat` with the change in the PR above I get epoch + 1 day:

```
root@8745ffb8cb0c:/# stat helioscanary-0.6.0.jar
  File: 'helioscanary-0.6.0.jar'
  Size: 32123778  	Blocks: 62744      IO Block: 4096   regular file
Device: 1dh/29d	Inode: 67          Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2015-06-10 04:51:50.673772456 +0000
Modify: 1970-01-02 00:00:00.000000000 +0000
Change: 2015-06-10 04:51:50.743771173 +0000
 Birth: -
```

**Comment by rohansingh** [comment]
I think the "right" way to do this is have the mtime of each file actually copied correctly from the source and not set to some arbitrary value. Let's say I write a program that reads the mtime of some assets and uses those in some way. If I use docker-maven-plugin, that will break since the mtime of the files copied into the image will not match the mtime of the source files on disk.

Main problem is that this is a pain in the ass or near-impossible to do with the way we're copying files using Maven, right?

**Comment by davidxia** [comment]
@rohansingh Would the PR above work?

[Javadocs say](http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#copy(java.nio.file.Path,%20java.nio.file.Path,%20java.nio.file.CopyOption...)):

> COPY_ATTRIBUTES	Attempts to copy the file attributes associated with this file to the target file. The exact file attributes that are copied is platform and file system dependent and therefore unspecified. Minimally, the last-modified-time is copied to the target file if supported by both the source and target file store. Copying of file timestamps may result in precision loss.

**Comment by davidxia** [comment]
@sjardine @creactiviti @chrisdostert @jleskovar please try out 0.2.11 and see if that works for you.

**Comment by jleskovar** [comment]
0.2.11 works fine for me now, cheers.

**Comment by davidxia** [comment]
Awesome

**Comment by creactiviti** [comment]
0.2.11 works great. thanks @davidxia

**Comment by lanoxx** [comment]
Hi, this has been fixed as of Docker 1.8 and so docker no longer takes mtime into account for the cache. Maybe you can drop this change and just use the original mtime of the file?

See: https://github.com/docker/docker/issues/4351
